### PR TITLE
add MaxText GRPO / RL post-training sample on GKE TPU v7x

### PR DESCRIPTION
**Pull request by pmotgi**
@## Description

This PR introduces a comprehensive new sample for running **MaxText GRPO (Reinforcement Learning) post-training** on Google Kubernetes Engine (GKE) using Cloud TPUs.

It provides a production-tested foundation for aligning Large Language Models (like Llama-3.1-8B-Instruct) using Group Relative Policy Optimization on TPU v7x infrastructure.

### Key Additions:
* **Kubernetes Manifests:** Includes both DWS-Flex and Spot JobSet manifests (`2x2x1` TPU v7x topology) and Kueue cluster queue configuration.
* **Architecture & Docker Deployment:** A dedicated runner image leveraging MaxText and vLLM (for rollout sampling) concurrently on the same TPU slice.
* **TRL to MaxText Mapping Guide:** A critical migration guide for users coming from HuggingFace TRL on GPUs. It explicitly details hyperparameter translation gotchas, such as how `batch_size` vs completions are counted, and sequence length padding behaviors to prevent HBM OOMs.
* **Customization Guides:** Includes a custom rewards template (`custom_rewards_template.py`) and a custom dataset integration guide.
* **MFU Calculation:** Includes a post-training MFU script (`calculate_mfu.py`) to measure hardware utilization.

co-contributor: @rick-c-goog

### CODE REVIEW

File: ai-ml/maxtext-rl-on-gke/docker/README.md [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+*(If using Podman, replace `docker` with `podman`).*
+
+### Option C: Fast Build from an Existing MaxText Runner Image (Optional)
+If your organization already maintains a base MaxText runner image, you can pass `--build-arg BASEIMAGE=...` to skip the initial setup step:
+
+```bash
+# Via Cloud Build:
+gcloud builds submit . \
+  --tag="${IMAGE_TAG}" \
+  --build-arg BASEIMAGE="YOUR_EXISTING_MAXTEXT_IMAGE" \
```
Comment (moficodes) [inline ai-ml/maxtext-rl-on-gke/docker/README.md]: Can we make the Baseimage a variable too?

File: ai-ml/maxtext-rl-on-gke/docker/README.md [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+              cloud.google.com/gke-tpu-accelerator: tpu7x
+              cloud.google.com/gke-tpu-topology: 2x2x1
+              cloud.google.com/gke-spot: 'true'
+            tolerations:
+            - key: google.com/tpu
+              operator: Exists
+              effect: NoSchedule
+            containers:
+            - name: lora-trainer
+              image: us-east5-docker.pkg.dev/YOUR_PROJECT/YOUR_REPO/maxtext-grpo-runner:v2
```
Comment (moficodes) [inline ai-ml/maxtext-rl-on-gke/docker/README.md]: should make this variables that gets replaces with something like envsubst. Or we should call out they need to replace this.

File: ai-ml/maxtext-rl-on-gke/CUSTOM_DATASET_AND_REWARDS_GUIDE.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+### Column Requirements:
+MaxText's built-in processor automatically reads:
+* **Prompt Column**: `prompt`, `question`, or `problem`
+* **Answer Column**: `answer`, `solution`, or `expected_answer`
+
+---
+
+## 5. Kubernetes JobSet Configuration
+
+Update lines 71–85 in your JobSet manifest ([`llama3.1-8b-grpo-dws-2x2x1-training.yaml`](file:///Users/pmotgi/exploration/cerence/training-grpo/llama3.1-8b-grpo-dws-2x2x1-training.yaml)):
```
Comment (moficodes) [inline ai-ml/maxtext-rl-on-gke/CUSTOM_DATASET_AND_REWARDS_GUIDE.md]: This is hardcoded to your filesystem.

File: ai-ml/maxtext-rl-on-gke/CUSTOM_DATASET_AND_REWARDS_GUIDE.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            volumeMounts:
+            - mountPath: /dev/shm
+              name: dshm
+            - mountPath: /checkpoint
+              name: gcs-input
+            - mountPath: /data                 # <-- Mount PVC containing dataset and custom_rewards.py
+              name: dataset-volume
+            containers:
+            - name: grpo-trainer
+              image: <YOUR_REGION>-docker.pkg.dev/<YOUR_PROJECT_ID>/<YOUR_REPO_NAME>/maxtext-grpo-runner:v2
```
Comment (moficodes) [inline ai-ml/maxtext-rl-on-gke/CUSTOM_DATASET_AND_REWARDS_GUIDE.md:156]: tell people to replace this values

File: ai-ml/maxtext-rl-on-gke/README.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
+# MaxText Post-Training RL / GRPO on TPU v7x
+
+This directory contains the production, tested, and verified Kubernetes JobSet manifest for running **Group Relative Policy Optimization (GRPO)** with **vLLM Rollout Sampling** on Google Cloud TPU v7x (`tpu-v7x-spot-2x2x1`).
+
+---
+
+## Verified Manifests
+
+### Option A: **Direct Spot Workload (`2x2x1` Topology)**
+* **Manifest File**: [`llama3.1-8b-grpo-spot-2x2x1-training.yaml`](file:///Users/pmotgi/exploration/cerence/training-grpo/llama3.1-8b-grpo-spot-2x2x1-training.yaml)
```
Comment (moficodes) [inline ai-ml/maxtext-rl-on-gke/README.md]: Update to relative path. This is your local filesystem.

File: ai-ml/maxtext-rl-on-gke/README.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+#### How to Run (Direct Spot):
+```bash
+kubectl apply -f training-grpo/llama3.1-8b-grpo-spot-2x2x1-training.yaml
+```
+
+---
+
+### Option B: **DWS-Flex & Kueue Orchestration (`2x2x1` Topology)**
+* **Kueue Setup File**: [`kueue-tpu7x-2x2x1-setup.yaml`](file:///Users/pmotgi/exploration/cerence/training-grpo/kueue-tpu7x-2x2x1-setup.yaml)
```
Comment (moficodes) [inline ai-ml/maxtext-rl-on-gke/README.md]: Update path.

File: ai-ml/maxtext-rl-on-gke/README.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+* **JobSet Manifest File**: [`llama3.1-8b-grpo-dws-2x2x1-training.yaml`](file:///Users/pmotgi/exploration/cerence/training-grpo/llama3.1-8b-grpo-dws-2x2x1-training.yaml)
```
Comment (moficodes) [inline ai-ml/maxtext-rl-on-gke/README.md]: Update path.

File: ai-ml/maxtext-rl-on-gke/README.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+#### 1. Create the DWS-Flex 2x2x1 Node Pool (Tested & Verified)
+```bash
+gcloud container node-pools create <CUSTOMER_NODEPOOL_NAME> \
```
Comment (moficodes) [inline ai-ml/maxtext-rl-on-gke/README.md]: Make the variables as a separate step so the whole command can be copied and pasted.

File: ai-ml/maxtext-rl-on-gke/README.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+---
+
+## Container Build Quickstart
+
+You can build the container using either **Google Cloud Build** or local **Docker / Podman**:
+
+### Option 1: Using Google Cloud Build
+```bash
+cd training-grpo/docker
+gcloud builds submit . --tag="REGION-docker.pkg.dev/PROJECT/REPO/maxtext-grpo-runner:v2" --machine-type="e2-highcpu-8"
```
Comment (moficodes) [inline ai-ml/maxtext-rl-on-gke/README.md]: use variables.

**Review by moficodes — CHANGES_REQUESTED** [review]
Few file paths are pointing to your local file system.

There are a lot of variable in yaml and commands that needs to be replaces manually by user. Look into using envsubst or let users know how and when to replace them.

File: ai-ml/maxtext-rl-on-gke/docker/README.md (hunk 1 above)
Comment (pmotgi) [reply ai-ml/maxtext-rl-on-gke/docker/README.md]: Yes, updated the change. Thanks!

File: ai-ml/maxtext-rl-on-gke/docker/README.md (hunk 2 above)
Comment (pmotgi) [reply ai-ml/maxtext-rl-on-gke/docker/README.md]: most often customers will have their own container here.  But I have added placeholder and readme so that they can update it as needed. Thanks

File: ai-ml/maxtext-rl-on-gke/CUSTOM_DATASET_AND_REWARDS_GUIDE.md (hunk 3 above)
Comment (pmotgi) [reply ai-ml/maxtext-rl-on-gke/CUSTOM_DATASET_AND_REWARDS_GUIDE.md]: good catch, thanks! I updated it.

File: ai-ml/maxtext-rl-on-gke/CUSTOM_DATASET_AND_REWARDS_GUIDE.md (hunk 4 above)
Comment (pmotgi) [reply ai-ml/maxtext-rl-on-gke/CUSTOM_DATASET_AND_REWARDS_GUIDE.md:156]: added a note section that mentions this. Thanks

File: ai-ml/maxtext-rl-on-gke/README.md (hunk 5 above)
Comment (pmotgi) [reply ai-ml/maxtext-rl-on-gke/README.md]: updated it. Thanks

File: ai-ml/maxtext-rl-on-gke/README.md (hunk 7 above)
Comment (pmotgi) [reply ai-ml/maxtext-rl-on-gke/README.md]: updated it. Thanks

File: ai-ml/maxtext-rl-on-gke/README.md (hunk 8 above)
Comment (pmotgi) [reply ai-ml/maxtext-rl-on-gke/README.md]: this would be a better approach. Thanks!

File: ai-ml/maxtext-rl-on-gke/README.md (hunk 9 above)
Comment (pmotgi) [reply ai-ml/maxtext-rl-on-gke/README.md]: updated it and added a note around it. Thanks

File: ai-ml/maxtext-rl-on-gke/README.md (hunk 6 above)
Comment (pmotgi) [reply ai-ml/maxtext-rl-on-gke/README.md]: done. Thanks

**Comment by pmotgi** [comment]
> Few file paths are pointing to your local file system.
> 
> There are a lot of variable in yaml and commands that needs to be replaces manually by user. Look into using envsubst or let users know how and when to replace them.

@moficodes - Thanks for your feedbacks. I have updated the PR with all the recommended changes. Let me know if there are any further changes needed. Thanks!

File: ai-ml/maxtext-rl-on-gke/docker/README.md [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+* **Sequence Length (`max_target_length`)**: `1024`
+* **Reward Functions**:
+  1. `match_format_exactly`: Enforces `<reasoning>...</reasoning><answer>...</answer>` tags.
+  2. `check_numbers`: Compares numerical values inside `<answer>` against GSM8K ground truth.
+
+---
+
+## 3. Production Kubernetes JobSet Manifest
+
+> **Note**: Replace placeholders (such as `<YOUR_REGION>`, `<YOUR_PROJECT_ID>`, `<YOUR_REPO_NAME>`, and `<YOUR_CHECKPOINT_PVC_NAME>`) with your specific cluster and Artifact Registry values before deploying. You can also template this manifest and substitute variables using tools like `envsubst` (e.g. `envsubst < jobset.yaml | kubectl apply -f -`).
```
Comment (moficodes) [inline ai-ml/maxtext-rl-on-gke/docker/README.md]: envsubst will only work if the yaml is in the form of $YOUR_REGION. If we are not giving the user the yaml file in that format envsubst will just confuse them.

File: ai-ml/maxtext-rl-on-gke/CUSTOM_DATASET_AND_REWARDS_GUIDE.md [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+* **Container Path (via PVC mount)**: `/checkpoint/scripts/custom_rewards.py` or `/data/custom_rewards.py`
+
+### Option B: Mount via Kubernetes ConfigMap (No GCS upload required)
+Create a Kubernetes ConfigMap containing your Python reward script:
+```bash
+kubectl create configmap grpo-custom-rewards \
+  --from-file=custom_rewards.py=./custom_rewards.py \
+  --namespace=default
+```
+And mount the ConfigMap in your JobSet:
```
Comment (moficodes) [inline ai-ml/maxtext-rl-on-gke/CUSTOM_DATASET_AND_REWARDS_GUIDE.md]: Where should they do it?

**Review by moficodes — CHANGES_REQUESTED** [review]
Couple more clarifications.

File: ai-ml/maxtext-rl-on-gke/CUSTOM_DATASET_AND_REWARDS_GUIDE.md (hunk 11 above)
Comment (pmotgi) [reply ai-ml/maxtext-rl-on-gke/CUSTOM_DATASET_AND_REWARDS_GUIDE.md]: This would go in any of the Jobset YAML file. I just added a sample part of how they can refer mountpath. Let me mention that it will be jobset yaml file

File: ai-ml/maxtext-rl-on-gke/docker/README.md (hunk 10 above)
Comment (pmotgi) [reply ai-ml/maxtext-rl-on-gke/docker/README.md]: Yeah let me remove the statement for envsubst since its not necessary. I think the previous statement clarifies that they just need to update the config as per their setup.

I will also add a note that container in this jobset is something that is built in previous step

File: ai-ml/maxtext-rl-on-gke/CUSTOM_DATASET_AND_REWARDS_GUIDE.md (hunk 11 above)
Comment (pmotgi) [reply ai-ml/maxtext-rl-on-gke/CUSTOM_DATASET_AND_REWARDS_GUIDE.md]: Since its just a reference- I dont think we need to provide an implementation for it, since the manifest below uses Option A (Copying the file to GCS bucket)

**Comment by pmotgi** [comment]
> Couple more clarifications.

no worries @moficodes - I just made few changes. PTAL and let me know if any changes are needed further. Thanks!

**Review by moficodes — APPROVED** [review]
lgtm
